src/org/python/modules/_locale/_locale.java

```diff
@@ -0,0 +1,718 @@
+        + "\n"
+        + "This function treats a missing encoding as using UTF-8.\n");
+
+    /**
+     *
+     * Java Locale loading behaviour is quite forgiving, or uninformative,
+     * depending on your perspective. It will return a dummy locale for any
+     * language tag that fits the syntax, or make various attempts to
+     * approximate a locale. This solution instead follows the stricter Python
+     * behaviour of requiring a particular locale to be installed.
+     */
+    public static PyString setlocale(PyInteger category, PyString locale) {
+        Py.writeDebug("_locale", "setlocale(category,locale==" + locale + ")");
+        if (locale == null || locale.equals(Py.None)
+                || locale.equals(currentLocale.getLocaleString())) {
+            return setlocale(category);
+        }
+        if (!category.equals(LC_ALL)) {
+            // Py.NotImplementedError not used as dependencies exist on
+            // locale.Error
+            throw LocaleException(
+                    "Only LC_ALL is supported in this version of Jython");
+        }
+        if (locale.equals(C_LOCALE_PY_STRING)) {
+            changeCurrentLocale(new CEmulationLocale());
+            return C_LOCALE_PY_STRING;
+        }
+        String localeStr = locale.toString();
+        String[] localeParts = localeStr.split("\\.");
+        String language;
+        String encoding = "UTF-8";
+        if (localeParts.length >= 3) {
+            throw LocaleException(
+                    "Does not conform to [language]_[locality].[encoding] format: "
+                            + locale.toString());
+        }
+        if (localeParts.length == 1) {
+            language = localeStr;
+        } else {
+            // == 2
+            language = localeParts[0];
+            encoding = localeParts[1];
+        }
+        Locale newLocale = loadLocale(language);
+        if (!isAvailableLocale(newLocale)) {
+            String underscoreAlias = language.replace('_', '-');
+            newLocale = loadLocale(underscoreAlias);
+            if (!isAvailableLocale(newLocale)) {
+                throw LocaleException("No locale found: " + locale.toString());
+            }
+        }
+        changeCurrentLocale(new JyLocale(newLocale, encoding));
+        return currentLocale.getLocaleString();
+    }
+
+    private static boolean isAvailableLocale(Locale locale) {
+        return locale != null && !locale.toString().isEmpty()
+                && AVAILABLE_LOCALES.contains(locale);
+    }
+
+    private static Locale loadLocale(String language) {
+        Locale newLocale = Locale.forLanguageTag(language);
+        Py.writeDebug("_locale",
+                "Current: " + currentLocale.getUnderlyingLocale());
+        Py.writeDebug("_locale",
+                "New: " + newLocale + " loaded with tag: " + language);
+        return newLocale;
+    }
+
+    private static synchronized void changeCurrentLocale(PyLocale pyLocale) {
```

Comment:
It is reasonable to synchronise writing this variable, but then you have to synchronise reading it too (and on the same monitor object, here the class). Otherwise, a consumer can legitimately hold a copy of the value.

Reply:
You are of course right, clearly this was underthought in the earlier patch. I have changed to volatile. As there are now only one stateful variable, usually changed only during initialization and testing under known concurrency conditions, it looks ok. Since it will be safe for most cases (and the cases where extra care is needed is now flagged in doco) I prefer to avoid the lock as it lowers risks of accidental deadlock. but happy to hear if there is a better way to align with the locking policy elsewhere. Or also if my concurrency reasoning is too rusty.